tests/Functional/DocumentManager/QueryBuilderTest.php

```diff
@@ -0,0 +1,272 @@
+        $issue3->setName('The hot water at ten and a closed car at four');
+        $issue3->setStatus('open');
+
+        $manager->persist($issue3, null, [
+            'path' => '/test/issue3',
+            'auto_create' => true,
+        ]);
+
+        $manager->flush();
+    }
+
+    /**
+     * It should query from an alias source.
+     */
+    public function testFromAlias()
+    {
+        $manager = $this->getDocumentManager();
+        $builder = $manager->createQueryBuilder();
+        $builder->setLocale('en');
+        $query = $builder->from()->document('full', 'p')->end()->getQuery();
+        $results = $query->execute();
+        $this->assertGreaterThan(1, count($results));
+    }
+
+    /**
+     * It should query from a class FQN source.
+     */
+    public function testFromDocumentClass()
+    {
+        $manager = $this->getDocumentManager();
+        $builder = $manager->createQueryBuilder();
+        $builder->setLocale('en');
+        $query = $builder->from()->document('Sulu\Component\DocumentManager\Tests\Functional\Model\FullDocument', 'p')->end()->getQuery();
+        $results = $query->execute();
+        $this->assertGreaterThan(1, count($results));
+    }
+
+    /**
+     * Test query by field.
+     */
+    public function testQueryByField()
+    {
+        $builder = $this->getDocumentManager()->createQueryBuilder();
+        $builder
+            ->setLocale('en')
+            ->from()
+                ->document('full', 'f')
+            ->end()
+            ->where()
+                ->eq()->field('f.title')->literal('Hello');
+
+        $results = $builder->getQuery()->execute();
+        $this->assertEquals(1, count($results));
+
+        return $builder;
+    }
+
+    /**
+     * It should take the locale into account when querying.
+     *
+     * @depends testQueryByField
+     */
+    public function testQueryByFieldWithLocale($builder)
+    {
+        $builder->setLocale('az');
+
+        $results = $builder->getQuery()->execute();
+        $this->assertEquals(0, count($results));
+    }
+
+    /**
+     * It should query on fields mapped by subscribers.
+     */
+    public function testQueryBySubscriberFields()
+    {
+        $oneHourAgo = new \DateTime();
+        $oneHourAgo->modify('-1 hour');
+        $builder = $this->getDocumentManager()->createQueryBuilder();
+        $builder
+            ->setLocale('en')
+            ->from()
+                ->document('full', 'f')
+            ->end()
+            ->where()
+                ->fieldIsset('f.created');
+
+        $results = $builder->getQuery()->execute();
+        $this->assertEquals(2, count($results));
+    }
+
+    /**
+     * It should join sources, hydrating objects from right().
+     */
+    public function testQueryBuilderFromJoinHydrateRight()
+    {
+        $builder = $this->getDocumentManager()->createQueryBuilder();
+        $builder->setLocale('en');
+        $builder->from('i')->joinInner()
```

Comment:
Basically no ..

The PHPCR-ODM initially used the Doctrine API, but it made it impossible
to express some of semantics of PHPCR.

I wrote [this document](https://gist.github.com/dantleech/298c9618b5889bddeb87)
before writing the PHPCR-ODM query builder which explains in more depth
the problems that were faced.

On Fri, Feb 19, 2016 at 12:47:31PM -0800, Bernhard Schussek wrote:
>    In [1]tests/Functional/DocumentManager/QueryBuilderTest.php:
> 
>  > +            ->end()
>  > +            ->where()
>  > +                ->fieldIsset('f.created');
>  > +
>  > +        $results = $builder->getQuery()->execute();
>  > +        $this->assertEquals(2, count($results));
>  > +    }
>  > +
>  > +    /**
>  > +     * It should join sources, hydrating objects from right().
>  > +     */
>  > +    public function testQueryBuilderFromJoinHydrateRight()
>  > +    {
>  > +        $builder = $this->getDocumentManager()->createQueryBuilder();
>  > +        $builder->setLocale('en');
>  > +        $builder->from('i')->joinInner()
> 
>    Can the query builder API be simplified to be more like Doctrine's query
>    builder API?
> 
>    —
>    Reply to this email directly or [2]view it on GitHub.
> 
>    Reverse link: [3]unknown
> 
> References
> 
>    Visible links
>    1. https://github.com/sulu-io/sulu-document-manager/pull/60#discussion_r53516058
>    2. https://github.com/sulu-io/sulu-document-manager/pull/60/files#r53516058
>    3. https://github.com/sulu-io/sulu-document-manager/pull/60/files#r53516058